### ISSUE TITLE
Bug in Saved Tensorflow model

### ISSUE THREAD

**Issue by Roshan1986** [tensorflow/tensorflow#41983]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Model training has unexplainable impact on older models saved on disk. I saved 1 tensorflow model(let's say 'best_model.h5') onto the disk which when loaded and evaluated gave me correct results. However, we I retrained my model which I didn't save to disk and then loaded my older model (i.e. the saved model 'best_model.h5'); the model evaluation results of saved model differed from the earlier load.
E.g with steps.
Step 1. Took UCI data for binary sentiment classification.
Step 2. Data Preparation -> Tokenize and Padded sequences of text.
Step 3. Simple NN architecture with Embedding, GlobalAveragePooling1D() and output Dense layer.
Step 4. Evaluated the model on Test data. Accuracy 83%
Step 5. Saved the model as 'best_model.h5'
Step 6. Loaded the 'best_model.h5' model and evaluated in Test Data. Accuracy 83%.
Step 7. Repeated all above steps except Step 5. Step 4 gave accuracy 81% and Step 6 gave accuracy 92%.
**Describe the expected behavior**
In the rerun as explained in Step 7 above, Step 6 should have maintained the accuracy of 83% instead it gave a different accuracy.
**Standalone code to reproduce the issue**
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

df=pd.read_csv('imdb_labelled.txt', sep='\t', header=None)
df.columns=['text', 'sentiment']

text=df.text.values
labels=df.sentiment.values

num_tokens=2000
max_len=20

tokenize=Tokenizer(num_words=num_tokens, oov_token='<OOV>')
tokenize.fit_on_texts(text)
sentences=tokenize.texts_to_sequences(text)

text_padded = pad_sequences(sentences, maxlen=max_len, padding='post', truncating='post')

X_train, X_test, y_train, y_test = train_test_split(text_padded, labels, test_size = 0.25)

initializer1 = tf.keras.initializers.GlorotUniform(seed=37)
initializer2 = tf.keras.initializers.GlorotNormal(seed=41)
model=tf.keras.Sequential([
                           tf.keras.layers.Embedding(num_tokens, 16, input_length=max_len, embeddings_initializer=initializer1),
                           tf.keras.layers.GlobalAveragePooling1D(),
                           tf.keras.layers.Dense(1, activation='sigmoid', kernel_initializer=initializer2)
])

model.summary()
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size=16, epochs=30, verbose=0)

y_pred=np.round(model.predict(X_test))

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

Model: "sequential_90"
Layer (type)                 Output Shape              Param 
embedding_91 (Embedding)     (None, 20, 16)            32000     
global_average_pooling1d_91  (None, 16)                0         
dense_90 (Dense)             (None, 1)                 17        
Total params: 32,017
Trainable params: 32,017
Non-trainable params: 0

Confusion matrix when trained again.
========================
[[65 20]
 [25 77]]
              precision    recall  f1-score   support

           0       0.72      0.76      0.74        85
           1       0.79      0.75      0.77       102

    accuracy                           0.76       187
   macro avg       0.76      0.76      0.76       187
weighted avg       0.76      0.76      0.76       187

0.7593582887700535

### Commented the model.save below to avoid overwrite of earlier saved model of 75.93%.
'# model.save('best_model.h5')
saved_model=tf.keras.models.load_model('best_model.h5')
saved_model.summary()
sm_y_pred=np.round(saved_model.predict(X_test))

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(confusion_matrix(y_test,sm_y_pred))
print(classification_report(y_test,sm_y_pred))
print(accuracy_score(y_test, sm_y_pred))

loss, acc = saved_model.evaluate(X_test,  y_test, verbose=2)
print('Restored model, accuracy: {:5.2f}%'.format(100*acc))

Model: "sequential"
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 20, 16)            32000     
global_average_pooling1d (Gl (None, 16)                0         
dense (Dense)                (None, 1)                 17        

Total params: 32,017
Trainable params: 32,017
Non-trainable params: 0

Confusion matrix of saved model.  Earlier the accuracy of saved model was 75.93% when initially saved.
==================================================================
[[89  7]
 [ 9 82]]
              precision    recall  f1-score   support

           0       0.91      0.93      0.92        96
           1       0.92      0.90      0.91        91

    accuracy                           0.91       187
   macro avg       0.91      0.91      0.91       187
weighted avg       0.91      0.91      0.91       187

0.9144385026737968
187/187 - 0s - loss: 0.3531 - accuracy: 0.9144
Restored model, accuracy: 91.44%

## The saved model predictions improved magically to 91.44%.

Attaching the dataset used by the code.
==========================
[imdb_labelled.txt](https://github.com/tensorflow/tensorflow/files/5012013/imdb_labelled.txt)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@Roshan1986 
please provide complete stand alone indented code to replicate the issue or a colab link with the issue reported for us to analyse. 
Can you please confirm if you have changed any parameters  or data when you say retrain

**Comment by Roshan1986** [comment]
I have provided the code and data in the issue description. Anyways, I am also providing the colab link https://colab.research.google.com/drive/1jpNMISMJGLILaCkhrj1NiEsqKCYLFBd0.

You will have to run the code as described in my issue description to understand the bug.

**Comment by Saduf2019** [comment]
@Roshan1986 
I do not have access to the gist shared.

**Comment by Roshan1986** [comment]
If you understood the bug.. you can try in any code of yours.. this bug doesn't necessarily require my code.. I believe it is a bug which can be noticed whenever you save your model and run the way I am running the code. Let me know if it is feasible at your end.

**Comment by gowthamkpr** [comment]
@Roshan1986 I have requested the access for your notebook. Please grant it. Thanks!

**Comment by jvishnuvardhan** [comment]
@Roshan1986 I followed your steps and I don't see any discrepancy in the evaluation results from the loaded model. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/995da4780c50e96fcf5d8c22e1a8492a/untitled21.ipynb).

As we don't have access to your code, I created one with `mnist` data. In future, try to share your code so that it will save some time for us and it will help faster resolution. Thanks!

Please verify once and close the issue if this was resolved for you. Thanks!
